designs/0003-parallel_autodiff.md

```diff
@@ -0,0 +1,701 @@
+- The new implementation for the old nested AD will be slower for
+  iterative things like ODE solving. The reason is that always using a
+  new independent AD tape requires to re-allocate it's needed memory
+  upon each iteration in the ODE integration process. However, if one
+  switches over to the more safe new pattern, then we retain the same
+  performance as we can then simply `recover()` the memory in the same
+  way as before which keeps the memory allocated, but ready for
+  reuse. Thus, the `coupled_ode_system` and the `algebraic_solver`
+  will need a light refactor to use the new scheme.
+- The suggested building blocks per se only parallelize the *forward*
+  sweep of reverse mode AD, but not the reverse sweep. Thus, the
+  operations which involved `adj_jac_apply`, for example, are not
+  speedup at all with this. However, this can easily be resolved by
+  introducing a general AD tree compression facility which works just
+  as `map_rect` does. Such a facility needs to know all the operands
+  and can then do nested AD and perform on the spot gradient
+  evaluation of the function of interest and store results as
+  `operands_and_partials`. From `map_rect` we have seen that this can
+  result in considerable speedups due to better use of CPU caches and
+  in addition, this approach will allow parallelize the forward and
+  the reverse sweep. However, such a facility can be written
+  independently of the RFC and can be used in combination with this
+  RFC. Doing so will be beneficial for a few reasons: The independent
+  AD tapes get compressed and we speedup things through keeping CPU
+  caches hot as only smaller AD tapes get generated and sweep-ed over.
+- We add the Intel TBB as a dependency if we go with this
+  approach. However, we could as a start try to limit its use to the
+  `parallel_map` and `parallel_reduce_sum` functions... but I would
+  actually opt against that and just use it like boost anywhere in
+  stan where it helps (for example, its malloc implementation speeds
+  up non-stiff ode integration on Mac by 25%).
+- The Intel TBB is developed by *Intel*, of course, and is only unit
+  tested on Intel hardware. However, it is still an open-source c++
+  library which will run fast on AMD as well. Moreover, the Intel TBB
+  forms the foundation for the parallel C++17 extensions for the clang
+  and gcc compilers.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+- Why is this design the best in the space of possible designs? I
+  think the approach is incurring minimal changes to Stan in order to
+  make it work with the TBB. The key is to refactor nesting and
+  introduce `ScopedChainableStack` in order to allow for local execution
+  contexts. The guiding principle is to have independent AD tapes used
+  in local contexts. This fits nicely into the TBB, but the introduced
+  abstraction is useful regardless of the TBB.
```

Comment:
Not sure there is any need to defend the choice of TBB but an added plus of Intel TBB is also that it works nicely with Intel MKL that we can already use to speedup Eigen. The same is true for OpenMP, but MKL+TBB is far superior - https://software.intel.com/en-us/articles/using-intel-mkl-and-intel-tbb-in-the-same-application.

This is a nice comparison on when one would chose TBB, OpenMP or native threading: https://software.intel.com/en-us/intel-threading-building-blocks-openmp-or-native-threads

One other option that is gaining momentum is SYCL ([example of a parallel for](https://www.codeplay.com/portal/sycl-tutorial-1-the-vector-addition)), but it suffers from the same issues as OpenMP in that it needs compiler extensions. And above all its still very very very young and doesnt have a lot of runtime support. 

My suggestion would be to extend this paragraph to cover all the alternatives and just write down why they were not chosen. There are going to be developers that come in and will ask "this is great but have you heard of XYZ" and we can just point them to this document. If you agree with my suggestion, I can help with this.

Reply:
The empirical evidence you point sounds good to include and I certainly don't see any reason why not to... so if you would like to add that, we can; sure.

We should certainly note that assessments from Intel are possibly biased; so that should be pointed out openly (although the docs you link is not any marketing gaga, of course).

What would you suggest to add the comments? Should I grant you permissions to the repo?

Reply:
I am going to make a PR on your fork.

Reply:
I think this is a great idea even if the intel docs are a little biased.